kotlin/custom-dictionary-style/src/main/java/com/esri/arcgisruntime/sample/customdictionarystyle/MainActivity.kt

```diff
@@ -0,0 +1,52 @@
+package com.esri.arcgisruntime.sample.customdictionarystyle
+
+import android.os.Bundle
+import androidx.appcompat.app.AppCompatActivity
+import com.esri.arcgisruntime.data.ServiceFeatureTable
+import com.esri.arcgisruntime.layers.FeatureLayer
+import com.esri.arcgisruntime.mapping.ArcGISMap
+import com.esri.arcgisruntime.mapping.Basemap
+import com.esri.arcgisruntime.mapping.Viewpoint
+import com.esri.arcgisruntime.symbology.DictionaryRenderer
+import com.esri.arcgisruntime.symbology.DictionarySymbolStyle
+import kotlinx.android.synthetic.main.activity_main.*
+
+class MainActivity : AppCompatActivity() {
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_main)
+
+    // create a new map with a streets basemap and set it to the map view
+    mapView.map = ArcGISMap(Basemap.createStreetsVector()).apply {
+      // create a feature layer from a service feature table
+      with(FeatureLayer(ServiceFeatureTable(getString(R.string.restaurants_url)))) {
+        // use a custom style to create a dictionary renderer and set it to the feature layer renderer
+        this.renderer = DictionaryRenderer(DictionarySymbolStyle.createFromFile(
+            getExternalFilesDir(getString(R.string.restaurant_stylx_path))?.path))
+        // once the feature layer has loaded
+        this.addDoneLoadingListener {
+          // set the map view's viewpoint to the feature layer extent
+          mapView.setViewpointAsync(Viewpoint(this.fullExtent))
+        }
+        // add the the feature layer to the map's operational layers
+        operationalLayers.add(this)
+      }
+    }
```

Comment:
I would prefer:
```kotlin
    // create a new map with a streets basemap and set it to the map view
    val map = ArcGISMap(Basemap.createStreetsVector())
    mapView.map = map

    // create a feature layer from a service feature table
    val featureTable = ServiceFeatureTable(getString(R.string.restaurants_url))
    val featureLayer = FeatureLayer(featureTable).apply {
      // use a custom style to create a dictionary renderer and set it to the feature layer renderer
      val dictionaryStyle = DictionarySymbolStyle.createFromFile(getExternalFilesDir(getString(R.string.restaurant_stylx_path))?.path)
      renderer = DictionaryRenderer(dictionaryStyle)
    }

    // set the map view's viewpoint to the feature layer extent when loaded
    featureLayer.addDoneLoadingListener { 
      mapView.setViewpointAsync(Viewpoint(featureLayer.fullExtent))
    }

    // add the the feature layer to the map's operational layers
    map.operationalLayers.add(featureLayer)
```

Use `apply` only in cases like this:
```kotlin
val myObject = MyObject("somethingA")
myObject.somethingC = ""
myObject.somethingF = ""
```
```kotlin
val myObject = MyObject("somethingA").apply {
  somethingC = ""
  somethingF = ""
}
```
It's just so that you don't have the redundant `myObject`. It's almost as if you're creating an overload constructor for `MyObject` that only takes the arguments you want to set. It replaces the builder pattern if you're familiar with that. Here, it would be nice if there was a FeatureLayer constructor that took a feature table and a renderer. The `apply` here nicely shows that you "built" a FeatureLayer using a required feature table with an "applied" renderer.

The mechanism that powers `with` is receiver type functions. One major use for those are DSLs. So you'll often do things with `with` that you'd maybe do with a DSL. For example, a SQL string builder:
```kotlin
val query = with (select("*")) {
   val fromStatement = if (something) {
      // Select.from
      from(MyTable)
   } else {
      from(AnotherTable)
  }

  with (fromStatement) {
    // From.where
    where(...)
  }
}
```

You can do a bunch of things *with* the scope of `select` to get your `from` and then do something *with* `from` and so on in a chaining manner. I wouldn't expect you to use it in samples much. Maybe if you were building a tree structure or a complex query based on a bunch of conditionals.

Reply:
Thanks for this @tschie - This is super useful for me trying to get the patterns of these uses right!

Reply:
Changed most things to fix your proposed patterns. I've stuck with doing the "map" related work all in the same place, because it still feels like building up the contents of the map with:

```
    // create a new map with a streets basemap and set it to the map view
    mapView.map =  ArcGISMap(Basemap.createStreetsVector()).apply {
      // add the the feature layer to the map's operational layers
      operationalLayers.add(featureLayer)
    }
```